File: backend/app/routes/tournaments.py
from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy.orm import Session
from app.database import get_db
from app.models import Tournament, Participant, Match, User, PlayerScore
from app.schemas import TournamentCreate, TournamentResponse, MatchCreate, MatchResponse, ScoreEntry
from app.security import get_current_user
from app.prize_service import PrizeService
from datetime import datetime

router = APIRouter(prefix="/api/tournaments", tags=["tournaments"])

@router.post("/", response_model=TournamentResponse)
def create_tournament(tournament: TournamentCreate, db: Session = Depends(get_db), current_user: User = Depends(get_current_user)):
    """Create a new tournament"""
    new_tournament = Tournament(
        title=tournament.title,
        description=tournament.description,
        organizer_id=current_user.id,
        entry_fee=tournament.entry_fee,
        max_participants=tournament.max_participants,
        game_mode=tournament.game_mode,
        prize_pool=tournament.entry_fee * tournament.max_participants * 0.85,
        start_date=tournament.start_date,
        registration_deadline=tournament.registration_deadline
    )
    
    db.add(new_tournament)
    db.commit()
    db.refresh(new_tournament)
    return new_tournament

@router.get("/{tournament_id}", response_model=TournamentResponse)
def get_tournament(tournament_id: int, db: Session = Depends(get_db)):
    """Get tournament details"""
    tournament = db.query(Tournament).filter(Tournament.id == tournament_id).first()
    if not tournament:
        raise HTTPException(status_code=404, detail="Tournament not found")
    return tournament

@router.get("/", response_model=list)
def list_tournaments(skip: int = 0, limit: int = 10, db: Session = Depends(get_db)):
    """List all tournaments"""
    tournaments = db.query(Tournament).offset(skip).limit(limit).all()
    return tournaments

@router.post("/{tournament_id}/matches", response_model=MatchResponse)
def create_match(tournament_id: int, match: MatchCreate, db: Session = Depends(get_db), current_user: User = Depends(get_current_user)):
    """Create a match for tournament"""
    tournament = db.query(Tournament).filter(Tournament.id == tournament_id).first()
    if not tournament:
        raise HTTPException(status_code=404, detail="Tournament not found")
    
    if tournament.organizer_id != current_user.id:
        raise HTTPException(status_code=403, detail="Not authorized")
    
    new_match = Match(
        tournament_id=tournament_id,
        match_number=match.match_number,
        game_mode=match.game_mode,
        room_id=match.room_id,
        room_password=match.room_password,
        scheduled_time=match.scheduled_time
    )
    
    db.add(new_match)
    db.commit()
    db.refresh(new_match)
    return new_match

@router.get("/{tournament_id}/matches")
def get_tournament_matches(tournament_id: int, db: Session = Depends(get_db)):
    """Get all matches for a tournament"""
    matches = db.query(Match).filter(Match.tournament_id == tournament_id).all()
    return matches

@router.post("/{tournament_id}/join")
def join_tournament(tournament_id: int, freefire_uid: str, db: Session = Depends(get_db), current_user: User = Depends(get_current_user)):
    """Join a tournament"""
    tournament = db.query(Tournament).filter(Tournament.id == tournament_id).first()
    
    if not tournament:
        raise HTTPException(status_code=404, detail="Tournament not found")
    
    if tournament.current_participants >= tournament.max_participants:
        raise HTTPException(status_code=400, detail="Tournament is full")
    
    existing = db.query(Participant).filter(
        Participant.tournament_id == tournament_id,
        Participant.user_id == current_user.id
    ).first()
    
    if existing:
        raise HTTPException(status_code=400, detail="Already registered")
    
    participant = Participant(
        user_id=current_user.id,
        tournament_id=tournament_id,
        freefire_uid=freefire_uid
    )
    
    db.add(participant)
    db.commit()
    db.refresh(participant)
    
    return participant

@router.get("/{tournament_id}/participants")
def get_participants(tournament_id: int, db: Session = Depends(get_db)):
    """Get all participants for a tournament"""
    participants = db.query(Participant).filter(
        Participant.tournament_id == tournament_id
    ).all()
    return participants

@router.post("/{tournament_id}/matches/{match_id}/scores")
def submit_score(tournament_id: int, match_id: int, score_data: ScoreEntry, db: Session = Depends(get_db), current_user: User = Depends(get_current_user)):
    """Submit player score for a match"""
    tournament = db.query(Tournament).filter(Tournament.id == tournament_id).first()
    
    if not tournament:
        raise HTTPException(status_code=404, detail="Tournament not found")
    
    if tournament.organizer_id != current_user.id:
        raise HTTPException(status_code=403, detail="Only organizer can submit scores")
    
    score = PrizeService.record_player_score(
        db=db,
        match_id=match_id,
        participant_id=score_data.participant_id,
        kills=score_data.kills,
        position=score_data.position,
        is_booyah=score_data.is_booyah
    )
    
    return score

@router.get("/{tournament_id}/leaderboard")
def get_leaderboard(tournament_id: int, db: Session = Depends(get_db)):
    """Get tournament leaderboard"""
    leaderboard = PrizeService.calculate_tournament_leaderboard(db, tournament_id)
    return leaderboard

@router.post("/{tournament_id}/matches/{match_id}/distribute-prizes")
def distribute_prizes(tournament_id: int, match_id: int, db: Session = Depends(get_db), current_user: User = Depends(get_current_user)):
    """Distribute prizes for a completed match"""
    tournament = db.query(Tournament).filter(Tournament.id == tournament_id).first()
    
    if not tournament:
        raise HTTPException(status_code=404, detail="Tournament not found")
    
    if tournament.organizer_id != current_user.id:
        raise HTTPException(status_code=403, detail="Not authorized")
    
    success = PrizeService.distribute_match_prizes(db, match_id)
    
    if not success:
        raise HTTPException(status_code=400, detail="Failed to distribute prizes")
    
    return {"message": "Prizes distributed successfully"}

File: backend/routes/tournaments.js
const express = require('express');
const router = express.Router();
const { authenticate } = require('../middleware/auth');
const db = require('../models');

// GET all tournaments
router.get('/', async (req, res) => {
  try {
    const tournaments = await db.Tournament.findAll({
      include: [
        {
          model: db.User,
          as: 'organizer',
          attributes: ['id', 'username', 'email'],
        },
        {
          model: db.TournamentParticipant,
          as: 'participants',
          attributes: ['id', 'user_id'],
        },
      ],
      order: [['created_at', 'DESC']],
    });

    res.json(tournaments);
  } catch (error) {
    console.error('Error fetching tournaments:', error);
    res.status(500).json({ error: 'Failed to fetch tournaments' });
  }
});

// GET single tournament
router.get('/:id', async (req, res) => {
  try {
    const { id } = req.params;

    const tournament = await db.Tournament.findByPk(id, {
      include: [
        {
          model: db.User,
          as: 'organizer',
          attributes: ['id', 'username', 'email'],
        },
        {
          model: db.TournamentParticipant,
          as: 'participants',
          include: [
            {
              model: db.User,
              attributes: ['id', 'username', 'email'],
            },
          ],
        },
      ],
    });

    if (!tournament) {
      return res.status(404).json({ error: 'Tournament not found' });
    }

    res.json(tournament);
  } catch (error) {
    console.error('Error fetching tournament:', error);
    res.status(500).json({ error: 'Failed to fetch tournament' });
  }
});

// POST create tournament
router.post('/', authenticate, async (req, res) => {
  try {
    const { name, description, game, max_participants, prize_pool, registration_deadline, start_date } = req.body;

    if (!name || !description || !game) {
      return res.status(400).json({ error: 'Missing required fields' });
    }

    const tournament = await db.Tournament.create({
      name,
      description,
      game,
      max_participants: max_participants || 16,
      prize_pool: prize_pool || 0,
      registration_deadline,
      start_date,
      organizer_id: req.user.id,
    });

    const result = await db.Tournament.findByPk(tournament.id, {
      include: [
        {
          model: db.User,
          as: 'organizer',
          attributes: ['id', 'username', 'email'],
        },
      ],
    });

    res.status(201).json(result);
  } catch (error) {
    console.error('Error creating tournament:', error);
    res.status(500).json({ error: 'Failed to create tournament' });
  }
});

// PUT update tournament
router.put('/:id', authenticate, async (req, res) => {
  try {
    const { id } = req.params;
    const { name, description, game, max_participants, prize_pool, status, registration_deadline, start_date } = req.body;

    const tournament = await db.Tournament.findByPk(id);

    if (!tournament) {
      return res.status(404).json({ error: 'Tournament not found' });
    }

    // Check if user is organizer
    if (tournament.organizer_id !== req.user.id) {
      return res.status(403).json({ error: 'Not authorized to update this tournament' });
    }

    await tournament.update({
      name: name || tournament.name,
      description: description || tournament.description,
      game: game || tournament.game,
      max_participants: max_participants || tournament.max_participants,
      prize_pool: prize_pool !== undefined ? prize_pool : tournament.prize_pool,
      status: status || tournament.status,
      registration_deadline: registration_deadline || tournament.registration_deadline,
      start_date: start_date || tournament.start_date,
    });

    const result = await db.Tournament.findByPk(id, {
      include: [
        {
          model: db.User,
          as: 'organizer',
          attributes: ['id', 'username', 'email'],
        },
      ],
    });

    res.json(result);
  } catch (error) {
    console.error('Error updating tournament:', error);
    res.status(500).json({ error: 'Failed to update tournament' });
  }
});

// DELETE tournament
router.delete('/:id', authenticate, async (req, res) => {
  try {
    const { id } = req.params;

    const tournament = await db.Tournament.findByPk(id);

    if (!tournament) {
      return res.status(404).json({ error: 'Tournament not found' });
    }

    // Check if user is organizer
    if (tournament.organizer_id !== req.user.id) {
      return res.status(403).json({ error: 'Not authorized to delete this tournament' });
    }

    await tournament.destroy();

    res.json({ message: 'Tournament deleted successfully' });
  } catch (error) {
    console.error('Error deleting tournament:', error);
    res.status(500).json({ error: 'Failed to delete tournament' });
  }
});

// GET tournament participants
router.get('/:id/participants', async (req, res) => {
  try {
    const { id } = req.params;

    const participants = await db.TournamentParticipant.findAll({
      where: { tournament_id: id },
      include: [
        {
          model: db.User,
          attributes: ['id', 'username', 'email'],
        },
      ],
    });

    res.json(participants);
  } catch (error) {
    console.error('Error fetching participants:', error);
    res.status(500).json({ error: 'Failed to fetch participants' });
  }
});

// POST register for tournament
router.post('/:id/register', authenticate, async (req, res) => {
  try {
    const { id } = req.params;
    const userId = req.user.id;

    const tournament = await db.Tournament.findByPk(id);

    if (!tournament) {
      return res.status(404).json({ error: 'Tournament not found' });
    }

    // Check if already registered
    const existing = await db.TournamentParticipant.findOne({
      where: { tournament_id: id, user_id: userId },
    });

    if (existing) {
      return res.status(400).json({ error: 'Already registered for this tournament' });
    }

    // Check if tournament is full
    const participantCount = await db.TournamentParticipant.count({
      where: { tournament_id: id },
    });

    if (participantCount >= tournament.max_participants) {
      return res.status(400).json({ error: 'Tournament is full' });
    }

    const participant = await db.TournamentParticipant.create({
      tournament_id: id,
      user_id: userId,
    });

    res.status(201).json(participant);
  } catch (error) {
    console.error('Error registering for tournament:', error);
    res.status(500).json({ error: 'Failed to register for tournament' });
  }
});

// DELETE unregister from tournament
router.delete('/:id/register', authenticate, async (req, res) => {
  try {
    const { id } = req.params;
    const userId = req.user.id;

    const participant = await db.TournamentParticipant.findOne({
      where: { tournament_id: id, user_id: userId },
    });

    if (!participant) {
      return res.status(404).json({ error: 'Not registered for this tournament' });
    }

    await participant.destroy();

    res.json({ message: 'Unregistered successfully' });
  } catch (error) {
    console.error('Error unregistering from tournament:', error);
    res.status(500).json({ error: 'Failed to unregister from tournament' });
  }
});

module.exports = router;
